### PR TITLE
Configurable CGO_ENABLED variable

### PR DESCRIPTION
**Pull request by jianzhangbjz**
This PR will fix the issue: #48420 . The default value of `CGO_ENABLED` is 0, it is configurable. We can set it via environment variable, the same as the `go env` shows(as the below). Hence, the will support the code of `cgo` compiling by modify the environment variable.
I think this will be more convenient and reasonable for ordinary users.
```
root@test1:/home/zhangjian/src/k8s.io/kubernetes# go env
GOARCH="amd64"
GOBIN="/usr/local/go/bin"
GOEXE=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOOS="linux"
GOPATH="/home/zhangjian"
GORACE=""
GOROOT="/usr/local/go"
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
CC="gcc"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build959694934=/tmp/go-build -gno-record-gcc-switches"
CXX="g++"
CGO_ENABLED="1"
```

### CODE REVIEW

**Comment by cblecker** [comment]
We should be careful here. I believe we disable CGO on purpose although I can't remember off the top of my head why.

Either way, I think we should either enable it or not.. not leave it ambiguous and pull in whatever the building user's environment is.

**Comment by spxtr** [comment]
/assign @ixdy 

> Either way, I think we should either enable it or not.. not leave it ambiguous and pull in whatever the building user's environment is.

+1 to that

**Comment by ixdy** [comment]
I think we set `CGO_ENABLED=0` because that was the only way to force these binaries to be statically linked.

I'm not sure this is still the case in modern versions of Go, however; notably, the bazel go rules don't set `CGO_ENABLED`, and we just pass `-linkmode external -extldflags -static`, which seems to have the desired effect.

I'm still searching to see if there was a particular Go version where the behavior changed.
cc @thockin @mikedanese @luxas

**Comment by luxas** [comment]
@ixdy I can't comment directly on that, since I think it's depending on how you installed Go. If you installed it to default to `CGO_ENABLED=0`, well, it'll use that by default. Not sure if that default has changed, it might have been like that.

Also worth noticing is that it might still change its mind and use cgo if you have deps like `net` and `os/user` or similar. The status as of v1.5/v1.6 is here: https://github.com/kubernetes/community/blob/master/contributors/design-proposals/multi-platform.md

**Comment by luxas** [comment]
That said, I see no reason to make this configurable really in this case.
A static binary is a static binary. Why would you want to make it (maybe) depend on libc by setting `CGO_ENABLED=1`?

**Comment by fejta** [comment]
Ditto everyone else's comments that we should turn it on or off, not make it configurable. Please add me back if I'm needed further.
/unassign

**Comment by jianzhangbjz** [comment]
@luxas Consider the following scenario, there are many `libxxx.so`(such as drivers, algorithms) made of the `c` code, however, not all nodes have these library files.  
In order to avoid running error, we need to dynamically load (call) these library files. So we have to use the `cgo`, hence need to enable `CGO_ENABLED`.
Base on the above scene, I think we need somewhere to turn it or not. Otherwise it would be a nightmare for the user, because users obviously enable the `CGO_ENABLED`, but would not take effect when compile.

**Comment by ixdy** [comment]
@jianzhangbjz the binaries listed in [`KUBE_STATIC_LIBRARIES`](https://github.com/kubernetes/kubernetes/blob/a9bf44101b3ec13a3e94df926c95ba5384d18e28/hack/lib/golang.sh#L195-L204) are copied into docker images and must be statically linked.

It may now be possible to statically link these binaries without disabling `cgo`, per my comments earlier.

**Comment by jianzhangbjz** [comment]
@ixdy Oh, I see, but why should we ensure that these libraries must be statically linked, dynamically linked is not more flexible? And if we want our `kube-scheduler` binary to link the `.so` files dynamically, how to do that?

**Comment by ixdy** [comment]
@jianglingxia We [add these binaries to docker images](https://github.com/kubernetes/kubernetes/blob/a9bf44101b3ec13a3e94df926c95ba5384d18e28/build/lib/release.sh#L324) (in particular, [these binaries](https://github.com/kubernetes/kubernetes/blob/a9bf44101b3ec13a3e94df926c95ba5384d18e28/build/common.sh#L94-L99)).
Because these use fairly lightweight base images, all dependencies must be compiled into the binaries statically.

**Comment by mikedanese** [comment]
If we want to add CGO to the scheduler in upstream kubernetes (which I'm not sure we do), we should remove kube-scheduler from KUBE_STATIC_LIBRARIES. What C code do you need to call from the scheduler?

**Comment by luxas** [comment]
Agree with @mikedanese and @ixdy. If that ever happens (which I don't see in the near future), we'll just have to move the component over to always being dynamically linked. But we don't even have to that, since it's possible to link all C libraries _into_ a statically Go library.

Eventually all output from Kubernetes might become statically linked (only the kubelet isn't, but it would technically be possible).

So I don't think it's worth it to make this configurable. The code that's being modified now is the case when the build _really always should compile statically_, so having some in-between state when you might get dynamically linked binaries for the things that should be static depending on your env doesn't cut it for me.

**Comment by jianzhangbjz** [comment]
@luxas @mikedanese It is our own algorithm library consisting of c code. Um... you mean, if we want to use it, we have to modify the [KUBE_STATIC_LIBRARIES](https://github.com/kubernetes/kubernetes/blob/a9bf44101b3ec13a3e94df926c95ba5384d18e28/hack/lib/golang.sh#L199) manually at present? Whether there is another way? Because the users do not know but also to modify [hack/lib/golang.sh](https://github.com/kubernetes/kubernetes/blob/a9bf44101b3ec13a3e94df926c95ba5384d18e28/hack/lib/golang.sh#L199), they only know that has enabled the `CGO_ENABLED`, but still errors being given, which will bring difficulties for their development.

**Comment by luxas** [comment]
@jianzhangbjz Please read https://github.com/kubernetes/community/blob/master/contributors/design-proposals/multi-platform.md (the reference section that touches Go compilation) if you haven't, you might get some more information from there.

TL;DR; We _might_ have to introduce a in-between case list: For binaries that reference external C libraries but should still be linked statically by linking the things it need into the binary.

@jianzhangbjz However, it sounds like you're building something that won't be merged into k8s core anyway? Why then use k8s' scripts?

**Comment by jianzhangbjz** [comment]
@luxas Yeah... our dynamical libraries which consisting of C code are separate from the `kube-scheduler` binary. Because we could not make sure all of the master node will install our own `libxxx.so` files. So, we need to call them dynamically.

**Comment by luxas** [comment]
> Yeah... our dynamical libraries which consisting of C code are separate from the kube-scheduler binary. Because we could not make sure all of the master node will install our own libxxx.so files. So, we need to call them dynamically.

So you're building a custom scheduler which needs external C libraries. That's fine.
But why would you need this patch then? Build your own scheduler with your own scripts ;)

**Comment by jianzhangbjz** [comment]
@luxas Yes, the external C libraries have been build with our own script. But, we also need to enable the `CGO_ENABLED` in k8s's makefile, because we added a new algorithm(calling the external C libs) in the `priority`. But we encountered errors, details in #48420 . We spent a lot of time to find the reason of errors.
 Finally, we found the problem was turned out to be caused by the `CGO_ENABLED`! But, we had enabled it before compile k8s, this is very confusing. So, I submitted this path to avoid modify the [hack/lib/golang.sh](https://github.com/kubernetes/kubernetes/blob/a9bf44101b3ec13a3e94df926c95ba5384d18e28/hack/lib/golang.sh) again and again manually and more easy to understand for other users who using `cgo` code.

**Comment by luxas** [comment]
> Yes, the external C libraries have been build with our own script. But, we also need to enable the CGO_ENABLED in k8s's makefile, because we added a new algorithm(calling the external C libs) in the priority. But we encountered errors, details in #48420 . We spent a lot of time to find the reason of errors.
Finally, we found the problem was turned out to be caused by the CGO_ENABLED! But, we had enabled it before compile k8s, this is very confusing. So, I submitted this path aim to avoid modify the hack/lib/golang.sh manually again and again and make it more easy to understand for other users who using cgo code.

Are that modifications to the core scheduler or to your-own-out-of-three thing?
Do you want to merge those changes into the core?

My spontaneous reaction is -- if we make the scheduler depend on C libraries, we should just remove it from the KUBE_STATIC_LIBRARIES list, not modify the behavior what happens to binaries in the KUBE_STATIC_LIBRARIES list

**Comment by jianzhangbjz** [comment]
Yes, the core scheduler. I see, thanks. :)

**Comment by jianzhangbjz** [comment]
@luxas It is better if we provide the instructions of KUBE_STATIC_LIBRARIES list in some docs, isn't it? Or Maybe I missed. So, should I close this PR?

**Comment by luxas** [comment]
@jianzhangbjz So -- I don't think a PR to add a dep on an external CGO library that would make the scheduler dynamically linked would be accepted as-is.

We must keep backwards-compability, so the scheduler **must always be statically linked**, no matter what.

The only way it _might_ work, is by making the scheduler use an external CGO lib, but link it **into** the binary at compile time.

Example:
```go
// main-cgo.go
package main
/*
char* sayhello(void) { return "Hello Kubernetes!"; }
*/
import "C"
import "fmt"
func main() {
	fmt.Println(C.GoString(C.sayhello()))
}
```
```bash
CGO_ENABLED=1 go build -ldflags '-extldflags "-static"' main-cgo.go
```

There is no list of binaries that does this currently, but we might want to create one if the need really exist.

**Comment by jianzhangbjz** [comment]
Okay, I see, closed it, thanks. :smiley:

**Comment by warmchang** [comment]
@jianzhangbjz ref: https://github.com/kubernetes/kubernetes/pull/64219

**Comment by jianzhangbjz** [comment]
@warmchang Thank you for your information!
